### PR TITLE
[SPARK-9403][SQL][WIP] codeGen in / inSet

### PR DESCRIPTION
**Pull request by tarekbecker**
Jira: https://issues.apache.org/jira/browse/SPARK-9403

@rxin ping

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 1]
```diff
@@ -107,21 +109,71 @@ case class In(value: Expression, list: Seq[Expression]) extends Predicate with C
...
+  }
+
 }
 
+/**
+ * Helper companion object in order to support code generation.
+ */
+object InSet {
+
+  @transient var hset: Set[Any] = null
```
Comment (tarekauel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: @rxin Is there a better way to expose `hset` to the codeGen stuff?

**Comment by rxin** [comment]
Can you beef up the test cases for in and inset in predicatesuites to include all primitive data types?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala (hunk 1 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: this won't work when you have multiple queries. lemme see ...

**Comment by rxin** [comment]
actually now i think about it more, in most cases it's probably a lot faster to just use the generated code for IN compared with a hashset lookup. In the case of a hashset lookup, it might make sense to change the OptimizeIn rule to only convert In to InSet if the number of instances is greater than say 50.

**Comment by rxin** [comment]
As for passing in the hashset, one way to do it is to relax the type constraint of references in codegenerator, and allow passing in arbitrary objects.

**Comment by rxin** [comment]
Jenkins, ok to test.

**Comment by rxin** [comment]
Jenkins, add to whitelist.

**Pull request closed without merging**